### ISSUE TITLE
atoms and error boundaries

### ISSUE THREAD

**Issue by JLarky** [pmndrs/jotai#92]
I'm trying out jotai in one of my projects and it was crashing with code like this
```
 React.useEffect(() => {
    setState(s => {
      const x = {};
      delete x[123][123];
    });
  });
 ```
and I was confused why this error wasn't caught by the react error boundary (of `react-error-boundary` fame). Then I realized that my component three was like this `JotaiProvider -> ErrorBoundary -> MyComponent` so doing something like `ErrorBoundary -> JotaiProvider -> ErrorBoundary -> MyComponent` helped, but it makes me realize interesting consideration that with smaller context providers I can have them closer to my component and thus making errors more local, but errors in jotai are basically globally scoped :)

I don't know if that's an issue per se, but maybe a good point to consider when deciding between custom context vs global context

**Comment by dai-shi** [comment]
ErrorBoundary should be inside JotaiProvider.
Provider should catch errors in the render phase and useAtom will throw it.
So, if what you observe is different, it's likely to be a bug.
Would you be willing to create a repro in codesandbox?

**Comment by JLarky** [comment]
OK :) that wasn't easy and I'm still not 100% there :) I managed to reproduce it basically by accident :)

So here are the steps:
1. go to https://codesandbox.io/s/jotai-issue-92-forked-uih92?file=/src/App.tsx
2. edit App.tsx source code and you will see
![image](https://user-images.githubusercontent.com/7026/95647231-a67a6f00-0a82-11eb-9861-f9e2cb298cb8.png)
and under overlay:
![image](https://user-images.githubusercontent.com/7026/95647236-b09c6d80-0a82-11eb-989f-533ac0fbc347.png)
3. expected behavior:
![image](https://user-images.githubusercontent.com/7026/95647251-d6297700-0a82-11eb-902a-fbcbaffcb0bc.png)
and under overlay:
![image](https://user-images.githubusercontent.com/7026/95647265-e7728380-0a82-11eb-9406-7088134ed51e.png)

I was trying to reproduce the issue with `unmount crash` button, which is closer to the code that I have in the real app, but I'm really struggling :) for me clicking on the button works as expected :) so the only way that I could manage to reproduce it was unmounting that is triggered by hot reload, but in my real app it has nothing to do with hot reload, and I see the same issue in production mode. So I will try to improve this code when I would have more time (I'm thinking maybe unmounting is async in the real app or just has more layers of components?), but I hope that the steps that I provided are enough to understand the issue :)

**Comment by dai-shi** [comment]
Thanks for trying out. tbh, we are still not very sure about hot reload behavior, so I downloaded your codesandbox, built and run locally.

It shows this after clicking the button.
```
EDIT this text in the App.tsx file to trigger hot reload and see bug #1
Expected oopsie... Now reload and try to trigger hot code reload instead
```

Is it the expected behavior?

**Comment by JLarky** [comment]
yes, that's how it suppose to function in all cases, but the bug is in the fact that sometimes it crashes on a higher level, and the only way I could reproduce it in codesandbox if you don't click the button at all, but trigger hot code reload instead (but as I said, in my real app it has nothing to do with code reload, I just couldn't find a way to reproduce it on codesandbox yet)

**Comment by dai-shi** [comment]
Hm, one note here is that the behavior of hot reload and real app can be caused by different reasons. I will investigate the hot reload behavior (which can be hard), but that may not fix the issue in the real app.

**Comment by dai-shi** [comment]
OK, I'd say the hot reload behavior is a limitation of current implementation (or theoretical one).
Check out this example:
https://codesandbox.io/s/jotai-issue-92-forked-om7e9?file=/src/atoms.ts

**Comment by JLarky** [comment]
> OK, I'd say the hot reload behavior is a limitation of current implementation (or theoretical one).
> Check out this example:
> https://codesandbox.io/s/jotai-issue-92-forked-om7e9?file=/src/atoms.ts

yes, this one works as expected. Did you just move atom into different file and that's it? :)

Anyway :) I will try to reproduce the exact issue I was having on my end if we can't do anything with this hot reload example :)

**Comment by dai-shi** [comment]
@JLarky Just let you know that the latest version 0.9.2 fixed a bug, which may or may not be relevant to your issue.

**Comment by JLarky** [comment]
I tried to reproduce it again, and as far as it looks both 0.7 and 0.9 work without issues. So I'm closing this because that thing with hot reload is not a huge deal, and I don't see those issues that I first saw.

**Comment by JLarky** [comment]
I take it back :)

https://codesandbox.io/s/jotai-issue-92-forked-mp5g2?file=/src/Crash.tsx

here's the code:
```
      setLoadAroundByRoom((x) => {});
      setGalleryFiles((x) => {
        if (1) throw new Error("");
      });
```

if you comment out one of those `set*` functions or change the order it will crash below JotaiProvider, but in this exact configuration it crashes higher than JotaiProvider

**Comment by dai-shi** [comment]
Thanks for the clarification!
OK, this is a bit controversial. Currently, it's by design.
We throw errors in render (atom.read) to error boundaries, but for atom.write we throw errors to the write function (even if it's async write).
What would be the behavior if you used `React.useState` instead of `useAtom`?
We should make it consistent as far as we could.

**Comment by JLarky** [comment]
@dai-shi what I don't understand in your reply is why it's different in this particular case? as I said if you run just one of those `set*` functions (so just comment out one of them) it crashed on the different level

**Comment by dai-shi** [comment]
So, with both `set*`, it shows `High level oopsie...`, and if we comment out the first one, it shows `Oopsie on Crash level...`. Good point.
 
This is because of the place we invoke atom.write function. But, I agree it's weird. I guess this reveals another issue about batched writes.

> We throw errors in render (atom.read) to error boundaries, but for atom.write we throw errors to the write function (even if it's async write).

I also find my explanation wasn't so accurate, because it shows error boundaries. I may have misunderstood something.

Let me think if I can write a test case for this.

**Comment by dai-shi** [comment]
> This is because of the place we invoke atom.write function. But, I agree it's weird. I guess this reveals another issue about batched writes.

This description might not be accurate either.

Anyway, the good news is I reproduced the issue in #141. The bad news is I don't know how to solve it yet.

**Comment by dai-shi** [comment]
@JLarky Hey, I think I fixed the issue in #141.
Would you try your example with the build? https://ci.codesandbox.io/status/pmndrs/jotai/pr/141/builds/63867

**Comment by dai-shi** [comment]
@JLarky Note it's already published as v0.10.0.

**Comment by JLarky** [comment]
can confirm that v0.10.0 fixes both example on codesandbox and the issue in my app
